analysis/README.md

```diff
@@ -0,0 +1,150 @@
+# Analysis of UUID usage in GitHub BigQuery Dataset
+
+## Background
+
+[RFC 4122](https://tools.ietf.org/html/rfc4122) defines three different classes of UUID versions
+with the following high level properties:
+
+- [`v4` are completely random](https://tools.ietf.org/html/rfc4122#section-4.4), very simple
+  algorithm, 122 random bits.
+- [`v1` are time- and MAC-address based](https://tools.ietf.org/html/rfc4122#section-4.2), rather
+  complex algorithm. If not used carefully, collisions are much more likely than with `v4`. `v1`
+  UUIDs are time-ordered.
+- [`v3` and `v5` are name-based](https://tools.ietf.org/html/rfc4122#section-4.3), special use
+  case. Constant input leads to constant output.
+
+Following the
+[principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) we
+assume that developers should always use the simplest UUID version that fulfills a given use case
+in order to reduces the risk of unexpected problems.
+
+- If all you need is a universally unique identifier you should always use `v4` UUIDs.
+- Only if you need time-ordering you should use `v1`.
+- Only if you need namespacing you should use `v3`/`v5`.
+
+In particular, accidentally using `v1` instead of `v4` UUIDs in cases where the developer is simply
+expecting a random value but is not aware of the fact that the generated IDs are time-ordered can
+have very negative consequences:
+
+- If these IDs are used as database keys and a database/cache does ID-based sharding (without
+  further hashing the shard keys), it can lead to "hot shards".
+- Developers who unintentionally use `v1` UUIDs in public datasets may not be aware of the fact
+  that the creation timestamp of the UUID and the MAC address of the computer that generates it may
+  be leaked (even though most modern implementations no longer leak the MAC address). See:
+  [“This privacy hole was used when locating the creator of the Melissa virus.”](https://en.wikipedia.org/wiki/Universally_unique_identifier#cite_note-11)
+
+We want to analyze current usage of the [`uuid` npm module](uuid-npm) in Open Source projects in
+order to design an appropriate API for the UUID standard library.
+
+## Hypothesis
+
+### Version Distribution
+
+**Hypothesis 1:** _`v4` is the by far most commonly used UUID version, followed by `v1` and only
+marginal amounts of `v3`/`v5` usage._
+
+We
+[expect that at least 80% - 90%](https://github.com/bcoe/proposal-standard-library-uuid/issues/3#issuecomment-489744827)
+of all `uuid` npm module usage exclusively makes use of `v4` UUIDs and that `v3`/`v5` usage is very
+uncommon (much less than 10%).
+
+**Consequence 1:** If above hypothesis can be validated we will only consider support `v4` UUIDs in
+the initial proposal.
+
+### Accidental `v1` Usage
+
+**Hypothesis 2:** _A considerable (=more than 50%) amount of `v1` UUID usage is "accidental" in the
```

Comment:
rather than saying `=more than 50%`, perhaps we can use a confidence interval, .e.g.,

`>50% of `v1` UUID usage is "accidental" with a confidence interval of 95%`

Reply:
OK. But how would you calculate the confidence interval (or rather standard deviation) in our case?

Reply:
My stats are rusty, so we should probably reach out for advice. But, I think, a binomial distribution is what we want. I believe it tells us what range our successes vs., failures fall into.

Reply:
I thought about this one again and I believe we're not on the right track here.

What we essentially want to figure out is whether _enough_ people use `v1` by accident (or because they think that `v1` must somehow be the "default") to justify that we promote `v4` UUIDs in a special defaulty way in our api.

So the question is, what is "enough" to convince us that choosing an interface that promotes `v4` as the default instead of an interface that is symmetric in the 4 versions is a _reasonable_ choice.

I believe specifying an exact threshold and a confidence interval creates pseudo-scientific expectations for a question that just cannot be answered in an exact, quantitative way. Even if we were able to exactly measure the ratio of projects that were using `v1` by accident (which would mean quite some additional effort) and even if we were able to quantify the confidence interval of this result, I doubt that this would help us answer the actual question.

I would argue that it doesn't really matter if 20% of repos use `v1` by accident or maybe even 90%: As long as the amount of accidental `v1` usage is not negligible it would be a strong argument for the asymmetric interface.

The reason why I arbitrarily put a `50%` there was that I assumed nobody would argue that "more than half" could be deemed _negligible_… But as stated above, I think even only 10% or 20% could still be enough…

So how about instead of trying to make this sound more precise (which IMO would be a lie anyways) let's make it sound more fuzzy:

> A non-negligible amount of `v1` UUID usage is "accidental" in the…

And then in the results argue that we consider 8 of the top 10 repos with accidental `v1` usage "non-negligible".

Reply:
@bcoe I've pushed a proposal, let me know what you think.

Reply:
at the end of the day, imagine 50% of people are using `v1` accidentally, all this really does is move us from 89% of people using V4 to 94% of people using V4 (or whatever).

Both numbers are pretty compelling, V4 is a pretty huge majority, and I think we should consider implementing it first (not that we're limiting ourselves from future additions).

Reply:
Please keep in mind that "Hypothesis 2" in my analysis is not so much concerned with the question of whether or not we should support anything but `v4`, that is already covered by "Hypothesis 1".

"Hypothesis 2" is about the question whether we want an interface that nudges users into using `v4` by default and that makes it artificially hard to choose `v1` or `v3`/`v5` "by accident" leaving it to those users who really need the more complex UUID versions (should we ever extend the interface to support those versions as well).